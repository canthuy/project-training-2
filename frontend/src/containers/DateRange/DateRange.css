.react-datepicker-wrapper {
  width: auto;
  margin-right: 10px;
}

.react-datepicker-wrapper input {
  background-color: #eeeeee;
  border-radius: 5px;
  border: none;
  padding: 5px;
  outline: none;
}
.btn-submit-date:active,.btn-submit-date:focus{
  box-shadow: none !important;
}
