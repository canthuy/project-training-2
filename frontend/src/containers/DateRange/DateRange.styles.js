import styled from 'styled-components';

const Label = styled.span`
  font-size: 16px;
  margin-left: 20px;
  margin-right: 5px;
`;
const DateRangeSC = styled.div`
  display: flex;
  align-items: center;
  margin-top: 10px;
`;
export { Label, DateRangeSC };
