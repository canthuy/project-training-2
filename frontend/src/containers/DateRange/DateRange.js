import DatePicker from 'react-datepicker';
import { useDispatch, useSelector } from 'react-redux';
import { useState } from 'react';
import { Button } from 'react-bootstrap';
import { getDate } from '../../redux/datepicker/actions';
import moment from 'moment';

import { Label, DateRangeSC } from './DateRange.styles';
import 'react-datepicker/dist/react-datepicker.css';
import './DateRange.css';

const DateRange = () => {
  const dispatch = useDispatch();
  const { startDate, endDate } = useSelector((state) => state.datepicker);
  const [dateRange, setDateRange] = useState([null, null]);
  const [newStartDate, newEndDate] = dateRange;
  console.log('startDate', startDate, 'endDate: ', endDate);
  const handleClick = () => {
    if (
      newStartDate !== null &&
      newEndDate !== null &&
      newStartDate !== startDate ||
      newEndDate !== endDate
    ) {
      dispatch(
        getDate({
          startDate: moment(newStartDate).format('DD/MM/YYYY'),
          endDate: moment(newEndDate).format('DD/MM/YYYY'),
        })
      );
    }
  };
  return (
    <DateRangeSC style={{ display: 'flex' }}>
      <Label>Range</Label>
      <DatePicker
        selectsRange={true}
        startDate={newStartDate}
        endDate={newEndDate}
        onChange={(update) => {
          setDateRange(update);
        }}
        dateFormat="dd/MM/yyyy"
      />
      <Button className="btn-submit-date" size="sm" onClick={handleClick}>
        Ok
      </Button>
    </DateRangeSC>
  );
};

export default DateRange;
