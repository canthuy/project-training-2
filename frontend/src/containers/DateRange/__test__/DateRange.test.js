import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { useSelector, useDispatch } from 'react-redux';
import DateRange from '../DateRange';
window.React = React;

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));
describe('DateRange component', () => {
  let mockState = {
    datepicker: {
      startDate: '',
      endDate: '',
    },
  };
  const mockDispatch = jest.fn();
  beforeEach(() => {
    useDispatch.mockImplementation(() => mockDispatch);
    useSelector.mockImplementation((callback) => {
      return callback(mockState);
    });
  });
  afterEach(() => {
    useDispatch.mockClear();
    useSelector.mockClear();
  });
  it('change date range and dispatch', () => {
    const { container, getByText, getByRole } = render(<DateRange />);
    getByRole('textbox').focus();
    const newSD = getByText('14');
    const newED = getByText('23');
    fireEvent.click(newSD);
    fireEvent.click(newED);
    expect(getByRole('textbox').value).toEqual(
      '14/01/2022 - 23/01/2022'
    );
    expect(container).toMatchSnapshot();
    fireEvent.click(getByRole('button', { name: /ok/i }));
    expect(mockDispatch).toHaveBeenCalled();
  });
  it('newStartDate, newEndDate is null', () => {
    const { getByRole } = render(<DateRange />);
    getByRole('textbox').focus();
    fireEvent.click(getByRole('button', { name: /ok/i }));
    expect(mockDispatch).not.toHaveBeenCalled();
  });
  it('startDate, endDate not null and do not change date range', () => {
    mockState = {
      datepicker: {
        startDate: '08/01/2022',
        endDate: '25/01/2022',
      },
    };
    const { getByText, getByRole } = render(<DateRange />);
    getByRole('textbox').focus();
    const newSD = getByText('8');
    const newED = getByText('25');
    fireEvent.click(newSD);
    fireEvent.click(newED);
    fireEvent.click(getByRole('button', { name: /ok/i }));
    expect(mockDispatch).not.toHaveBeenCalled();
  });
});
