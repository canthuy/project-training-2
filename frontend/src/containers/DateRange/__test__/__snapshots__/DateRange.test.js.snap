// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateRange component change date range and dispatch 1`] = `
<div>
  <div
    class="sc-gsDKAQ hGZSxA"
    style="display: flex;"
  >
    <span
      class="sc-bdvvtL clXdJo"
    >
      Range
    </span>
    <div
      class="react-datepicker-wrapper"
    >
      <div
        class="react-datepicker__input-container"
      >
        <input
          class=""
          type="text"
          value="14/01/2022 - 23/01/2022"
        />
      </div>
    </div>
    <button
      class="btn-submit-date btn btn-primary btn-sm"
      type="button"
    >
      Ok
    </button>
  </div>
</div>
`;
